indexing-service/src/main/java/org/apache/druid/indexing/worker/http/ShuffleResource.java

```diff
@@ -0,0 +1,116 @@
+ * under the License.
+ */
+
+package org.apache.druid.indexing.worker.http;
+
+import com.google.common.io.ByteStreams;
+import com.google.inject.Inject;
+import com.sun.jersey.spi.container.ResourceFilters;
+import org.apache.druid.indexing.worker.IntermediaryDataManager;
+import org.apache.druid.java.util.common.DateTimes;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.server.http.security.TaskShuffleResourceFilter;
+import org.joda.time.Interval;
+
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status;
+import javax.ws.rs.core.StreamingOutput;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.List;
+
+@Path("/druid/worker/v1/shuffle")
+@ResourceFilters(TaskShuffleResourceFilter.class)
+public class ShuffleResource
+{
+  private static final Logger log = new Logger(ShuffleResource.class);
+
+  private final IntermediaryDataManager intermediaryDataManager;
+
+  @Inject
+  public ShuffleResource(IntermediaryDataManager intermediaryDataManager)
+  {
+    this.intermediaryDataManager = intermediaryDataManager;
+  }
+
+  @GET
+  @Path("/task/{supervisorTaskId}/partition")
+  @Produces(MediaType.APPLICATION_OCTET_STREAM)
+  public Response getPartition(
+      @PathParam("supervisorTaskId") String supervisorTaskId,
+      @QueryParam("dataSource") String dataSource,
+      @QueryParam("startTime") String startTime,
+      @QueryParam("endTime") String endTime,
+      @QueryParam("partitionId") int partitionId
+  )
+  {
+    final Interval interval = new Interval(DateTimes.of(startTime), DateTimes.of(endTime));
+    final List<File> partitionFiles = intermediaryDataManager.findPartitionFiles(
+        supervisorTaskId,
+        interval,
+        partitionId
+    );
+
+    if (partitionFiles.isEmpty()) {
+      final String errorMessage = StringUtils.format(
+          "Can't find the partition for supervisor[%s], interval[%s], and partitionId[%s]",
+          supervisorTaskId,
+          interval,
+          partitionId
+      );
+      return Response.status(Status.NOT_FOUND).entity(errorMessage).build();
+    } else {
+      return Response.ok(
+          (StreamingOutput) output -> {
+            for (File partitionFile : partitionFiles) {
+              try (final FileInputStream fileInputStream = new FileInputStream(partitionFile)) {
+                ByteStreams.copy(fileInputStream, output);
+              }
+            }
+          }
+      ).build();
+    }
+  }
+
+  @DELETE
+  @Path("/task/{supervisorTaskId}")
+  public Response deletePartitions(
+      @PathParam("supervisorTaskId") String supervisorTaskId,
+      @QueryParam("dataSource") String dataSource
```

Comment:
unused ?

Reply:
ok... but why does it need to be declared here ? I understand that url needs that in query param for the filter to work but we probably don't need to capture that param in here.

Reply:
Hmm, I just removed dataSource and changed to use `StateResourceFilter` instead. Added the below javadoc.

> We use StateResourceFilter here because it performs an admin-like authorization and all endpoints here are supposed to be used for only internal communcation. Another possible alternate could be performing datasource-level authorization as in TaskResourceFilter. However, datasource information is not available in middleManagers or indexers yet which makes hard to use it. We could develop a new ResourceFileter in the future if needed.